### PR TITLE
Feature/customable authorized users only error response

### PR DESCRIPTION
**Pull request by abeyuya**
I have added new methods for customize error response when "Authorized users only".

### CODE REVIEW

File: lib/devise_token_auth/controllers/helpers.rb [hunk 1]
```diff
@@ -90,14 +95,15 @@ def log_process_action(payload)
...
-      #     admin_session       # Session data available only to the admin scope
+      #     authenticate_user!                   # Signs user in or 401
+      #     authenticate_admin!                  # Signs admin in or 401
+      #     user_signed_in?                      # Checks whether there is a user signed in or not
+      #     admin_signed_in?                     # Checks whether there is an admin signed in or not
+      #     current_user                         # Current signed in user
+      #     current_admin                        # Current signed in admin
+      #     user_session                         # Session data available only to the user scope
+      #     admin_session                        # Session data available only to the admin scope
+      #     render_authenticate_#{mapping}_error # Render error if
```
Comment (olleolleolle) [inline lib/devise_token_auth/controllers/helpers.rb]: Perhaps this docs line needs to be 2 lines (one for each generated mapping).

And the comment at the end of the line seems unfinished.

Comment (olleolleolle) [inline lib/devise_token_auth/controllers/helpers.rb]: I mean: have two lines, with ending comments like these:

- `# Render error unless user is signed in`
- `# Render error unless admin is signed in`

Comment (abeyuya) [inline lib/devise_token_auth/controllers/helpers.rb]: @olleolleolle Thank you for your review.
You are right. I will modify the comment soom.

Comment (olleolleolle) [inline lib/devise_token_auth/controllers/helpers.rb]: _That was fast._

**Comment by abeyuya** [comment]
When will this PR be merged? 🤔

**Comment by booleanbetrayal** [comment]
@abeyuya - There are conflicts and there is also an issue where the error response is not pluralized to match the prior string. Ie - "user" vs "users" (expected).

File: lib/devise_token_auth/controllers/helpers.rb [hunk 2]
```diff
@@ -68,7 +68,7 @@ def current_#{group_name.to_s.pluralize}
 
             def render_authenticate_#{group_name}_error
               return render json: {
-                errors: ["Authorized #{group_name} only."]
+                errors: [I18n.t('devise.failure.unauthenticated')]
```
Comment (abeyuya) [inline lib/devise_token_auth/controllers/helpers.rb]: I have resolved conflict about error message.

File: test/controllers/demo_group_controller_test.rb [hunk 3]
```diff
@@ -76,15 +76,15 @@ class DemoGroupControllerTest < ActionDispatch::IntegrationTest
...
           end
 
           it 'should define render_authenticate_mang_error' do
-            assert @controller.methods.include?(:render_authenticate_mang_error)
+            assert @controller.methods.include?(:render_authenticate_error)
           end
 
           it 'should define render_authenticate_member_error' do
-            assert @controller.methods.include?(:render_authenticate_member_error)
+            assert @controller.methods.include?(:render_authenticate_error)
```
Comment (olleolleolle) [inline test/controllers/demo_group_controller_test.rb]: These three tests now make less sense: their `it` descriptions talk about method names that are not checked for existence.

File: lib/devise_token_auth/controllers/helpers.rb [hunk 4]
```diff
@@ -66,14 +66,14 @@ def current_#{group_name.to_s.pluralize}
               end.compact
             end
 
-            def render_authenticate_#{group_name}_error
+            def render_authenticate_error
```
Comment (abeyuya) [inline lib/devise_token_auth/controllers/helpers.rb]: I have changed the method name to `render_authenticate_error` .

Exists error-render methods are using simple method name like `render_create_error`.
They don't include #{group_name} and #{mapping} like `render_create_error`.
So new method name also don't include these.

https://github.com/lynndylanhurley/devise_token_auth#registrations-controller

File: test/controllers/demo_mang_controller_test.rb [hunk 5]
```diff
@@ -48,11 +48,11 @@ class DemoMangControllerTest < ActionDispatch::IntegrationTest
           end
 
           it 'should define render_authenticate_user_error' do
-            assert @controller.methods.include?(:render_authenticate_user_error)
+            assert @controller.methods.include?(:render_authenticate_error)
           end
 
           it 'should define render_authenticate_mang_error' do
-            assert @controller.methods.include?(:render_authenticate_mang_error)
+            assert @controller.methods.include?(:render_authenticate_error)
```
Comment (olleolleolle) [inline test/controllers/demo_mang_controller_test.rb]: Also here: The same note about "tests' contents do not match their descriptions".

File: test/controllers/demo_user_controller_test.rb [hunk 6]
```diff
@@ -49,11 +49,11 @@ class DemoUserControllerTest < ActionDispatch::IntegrationTest
           end
 
           it 'should define render_authenticate_user_error' do
-            assert @controller.methods.include?(:render_authenticate_user_error)
+            assert @controller.methods.include?(:render_authenticate_error)
           end
 
           it 'should define render_authenticate_mang_error' do
-            assert @controller.methods.include?(:render_authenticate_mang_error)
+            assert @controller.methods.include?(:render_authenticate_error)
```
Comment (olleolleolle) [inline test/controllers/demo_user_controller_test.rb]: (And here, too.)

Perhaps there's more investigation to be made?

Comment (olleolleolle) [inline test/controllers/demo_user_controller_test.rb]: (Or, rather, renaming some tests?)

Comment (abeyuya) [inline test/controllers/demo_user_controller_test.rb]: Oh sorry 😱 
I will modify bad tests.

**Comment by zachfeldman** [comment]
@booleanbetrayal have a moment to finish review of this PR? All tests pass and no conflicts :)

**Comment by zachfeldman** [comment]
Going to merge it then!

**Pull request merged**